File: ch8/my_cpp_pkg/src/reset_counter_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "my_robot_interfaces/srv/reset_counter.hpp"

using ResetCounter = my_robot_interfaces::srv::ResetCounter;
using namespace std::chrono_literals;
using namespace std::placeholders;


class ResetCounterClientNode : public rclcpp::Node
{
public:
    ResetCounterClientNode() : Node("reset_counter_client")
    {
        client_ = this->create_client<ResetCounter>("reset_counter");
    }

void callResetCounter(int value)
{
    while (!client_->wait_for_service(1s)) {
        RCLCPP_WARN(this->get_logger(), "Waiting for the server...");
    }
    auto request = std::make_shared<ResetCounter::Request>();
    request->reset_value = value;
    client_->async_send_request(
        request, std::bind(&ResetCounterClientNode::callbackResetCounterResponse, this, _1));
}

private:
    void callbackResetCounterResponse(rclcpp::Client<ResetCounter>::SharedFuture future)
    {
        auto response = future.get();
        RCLCPP_INFO(this->get_logger(), "Success flag: %d, Message: %s", (int)response->success, response->message.c_str());
    }

    rclcpp::Client<ResetCounter>::SharedPtr client_;

};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<ResetCounterClientNode>();
    node->callResetCounter(20);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ch7/my_py_pkg/my_py_pkg/reset_counter_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from my_robot_interfaces.srv import ResetCounter


class ResetCounterClientNode(Node):
    def __init__(self):
        super().__init__("reset_counter_client")
        self.client_ = self.create_client(ResetCounter, "reset_counter")

    def call_reset_counter(self, value):
        while not self.client_.wait_for_service(1.0):
            self.get_logger().warn("Waiting for service...")
        request = ResetCounter.Request()
        request.reset_value = value
        future = self.client_.call_async(request)
        future.add_done_callback(self.callback_reset_counter_response)

    def callback_reset_counter_response(self, future):
        response = future.result()
        self.get_logger().info("Success flag: " + str(response.success))
        self.get_logger().info("Message: " + str(response.message))
 

def main(args=None):
    rclpy.init(args=args)
    node = ResetCounterClientNode()
    node.call_reset_counter(20)
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
